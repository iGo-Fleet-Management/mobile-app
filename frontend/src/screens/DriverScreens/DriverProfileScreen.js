import React, { useState, useEffect } from 'react';
import { View, Text, StyleSheet, TouchableOpacity, ScrollView, ActivityIndicator } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { MaterialIcons } from '@expo/vector-icons';
import LogoutConfirmation from '../../components/common/Logout';
import { API_IGO } from '@env';
import AsyncStorage from '@react-native-async-storage/async-storage';

import ProfileSkeleton from '../../components/common/Skeleton';

const DriverProfileScreen = ({ navigation }) => {
  const [userData, setUserData] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const [showLogoutModal, setShowLogoutModal] = useState(false);

  useEffect(() => {
    fetchUserProfile();
  }, []);

  const fetchUserProfile = async () => {
    try {
      setLoading(true);
      setError(null);
      // Get the authentication token from AsyncStorage
      const token = await AsyncStorage.getItem('userToken');
      
      if (!token) {
        throw new Error('No authentication token found');
      }
  
      const response = await fetch(`${API_IGO}/profile`, {
        method: 'GET',
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
      });
  
      if (!response.ok) {
        throw new Error('Failed to fetch profile data');
      }
  
      const responseData = await response.json();
      
      // Extract the user data from the nested structure
      if (responseData.success && responseData.data) {
        setUserData(responseData.data);
      } else {
        throw new Error('Invalid data format from API');
      }
      
      setLoading(false);
    } catch (err) {
      console.error('Error fetching profile:', err);
      setError(err.message);
      setLoading(false);
    }
  };

  const handleEditProfile = () => {
    navigation.navigate('EditProfile', { userData });
  };

  const handleAddressesPress = () => {
    navigation.navigate('EditAddresses', { addresses: userData?.addresses || [] });
  };

  const handleLogout = () => {
    setShowLogoutModal(true);
  };

  const confirmLogout = async () => {
    try {
      await AsyncStorage.removeItem('userToken');
      setShowLogoutModal(false);
      navigation.navigate('Login');
    } catch (err) {
      console.error('Error during logout:', err);
    }
  };

  const cancelLogout = () => {
    setShowLogoutModal(false);
  };

  // Render the header (shared between all states)
  const renderHeader = () => (
    <View style={styles.header}>
      <TouchableOpacity onPress={() => navigation.goBack()} style={styles.backButton}>
        <MaterialIcons name="chevron-left" size={30} color="black" />
      </TouchableOpacity>
      <TouchableOpacity onPress={() => {}} style={styles.helpButton}>
        <Text style={styles.helpText}>Ajuda</Text>
        <MaterialIcons name="help-outline" size={20} color="#007BFF" />
      </TouchableOpacity>
    </View>
  );

  // Render profile content or error message
  const renderContent = () => {
    if (loading) {
      return <ProfileSkeleton />;
    }

    if (error) {
      return (
        <View style={styles.errorContainer}>
          <MaterialIcons name="error-outline" size={60} color="#e53935" />
          <Text style={styles.errorText}>Erro ao carregar perfil: {error}</Text>
          <TouchableOpacity style={styles.retryButton} onPress={fetchUserProfile}>
            <Text style={styles.retryButtonText}>Tentar novamente</Text>
          </TouchableOpacity>
        </View>
      );
    }

    return userData ? (
      <>
        <View style={styles.profileHeader}>
          <View style={styles.avatarContainer}>
            <View style={styles.avatar}>
              <MaterialIcons name="person" size={60} color="black" />
            </View>
          </View>
          <Text style={styles.userName}>{`${userData.name} ${userData.last_name}`}</Text>
          <Text style={styles.userRole}>Motorista</Text>
        </View>

        <View style={styles.infoSection}>
          {userData.phone && (
            <>
              <Text style={styles.infoLabel}>Telefone:</Text>
              <Text style={styles.infoValue}>{userData.phone}</Text>
            </>
          )}
          
          {userData.addresses && userData.addresses.length > 0 && (
            <>
              <Text style={styles.infoLabel}>Endere√ßos:</Text>
              {userData.addresses.map((endereco, index) => (
                <Text key={index} style={styles.addressText}>
                  ({endereco.address_type}) {endereco.street} {endereco.number}, {endereco.neighbourhood}, {endereco.city}
                </Text>
              ))}
            </>
          )}
          
          <Text style={styles.infoLabel}>Email:</Text>
          <Text style={styles.infoValue}>{userData.email}</Text>
        </View>
      </>
    ) : null;
  };

  // Always show action buttons (Edit and Logout) regardless of content state
  const renderActionButtons = () => (
    <>
      <TouchableOpacity 
        style={styles.editButton} 
        onPress={handleEditProfile}
      >
        <Text style={styles.editButtonText}>Editar Perfil</Text>
      </TouchableOpacity>

      <TouchableOpacity 
        style={styles.logoutButton} 
        onPress={handleLogout}
      >
        <Text style={styles.logoutButtonText}>Sair</Text>
      </TouchableOpacity>
    </>
  );

  return (
    <SafeAreaView style={styles.container} edges={['top']}>
      {renderHeader()}
      
      <ScrollView style={styles.content}>
        {renderContent()}
        {renderActionButtons()}
      </ScrollView>

      <LogoutConfirmation
        visible={showLogoutModal}
        onConfirm={confirmLogout}
        onCancel={cancelLogout}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  content: {
    flex: 1,
  },
  errorContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    padding: 20,
    marginTop: 50,
    marginBottom: 20,
  },
  errorText: {
    fontSize: 16,
    color: '#e53935',
    textAlign: 'center',
    marginTop: 10,
    marginBottom: 20,
  },
  retryButton: {
    backgroundColor: '#4285F4',
    paddingVertical: 12,
    paddingHorizontal: 20,
    borderRadius: 25,
  },
  retryButtonText: {
    color: '#fff',
    fontSize: 16,
    fontWeight: 'bold',
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingHorizontal: 15,
    height: 50,
  },
  backButton: {
    padding: 5,
  },
  helpButton: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  helpText: {
    color: '#007BFF',
    marginRight: 5,
  },
  profileHeader: {
    alignItems: 'center',
    marginVertical: 20,
  },
  avatarContainer: {
    width: 100,
    height: 100,
    borderRadius: 50,
    backgroundColor: '#f5f5f5',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 15,
  },
  avatar: {
    width: 100,
    height: 100,
    borderRadius: 50,
    justifyContent: 'center',
    alignItems: 'center',
  },
  userName: {
    fontSize: 22,
    fontWeight: 'bold',
  },
  userRole: {
    fontSize: 16,
    color: '#666',
    marginTop: 5,
  },
  infoSection: {
    paddingHorizontal: 20,
    marginBottom: 20,
  },
  infoLabel: {
    fontSize: 16,
    fontWeight: 'bold',
    marginTop: 15,
    marginBottom: 5,
  },
  infoValue: {
    fontSize: 16,
    color: '#333',
  },
  addressText: {
    fontSize: 16,
    color: '#333',
    marginBottom: 5,
  },
  editButton: {
    backgroundColor: 'white',
    borderWidth: 1,
    borderColor: '#ddd',
    borderRadius: 25,
    paddingVertical: 12,
    paddingHorizontal: 20,
    marginHorizontal: 50,
    marginBottom: 15,
    alignItems: 'center',
  },
  editButtonText: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  logoutButton: {
    backgroundColor: 'white',
    borderWidth: 1,
    borderColor: '#ddd',
    borderRadius: 25,
    paddingVertical: 12,
    paddingHorizontal: 20,
    marginHorizontal: 50,
    marginBottom: 30,
    alignItems: 'center',
  },
  logoutButtonText: {
    fontSize: 16,
    fontWeight: 'bold',
    color: 'red',
  }
});

export default DriverProfileScreen;