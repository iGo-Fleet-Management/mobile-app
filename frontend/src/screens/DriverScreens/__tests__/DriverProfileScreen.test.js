import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import DriverProfileScreen from '../DriverProfileScreen'; 
// Mock dos ícones e dependências
jest.mock('@expo/vector-icons', () => ({
  MaterialIcons: ({ name }) => {
    const { Text } = require('react-native');
    return <Text>{name}</Text>; // Renderiza o nome do ícone como texto
  },
}));

jest.mock('react-native-safe-area-context', () => ({
  SafeAreaView: ({ children }) => children,
  useSafeAreaInsets: () => ({ top: 0, right: 0, bottom: 0, left: 0 })
}));

describe('DriverProfileScreen', () => {
  const mockNavigation = {
    navigate: jest.fn(),
    goBack: jest.fn(),
    setParams: jest.fn()
  };

  const mockUserData = {
    nome: 'John',
    sobrenome: 'Doe',
    telefone: '(31) 9 1234-5678',
    email: 'johndoe@gmail.com',
    enderecos: [
      { tipo: 'Casa', logradouro: 'R.', numero: '12', bairro: 'Bairro 1', cidade: 'Ipatinga' },
      { tipo: 'Trabalho', logradouro: 'R.', numero: '72', bairro: 'Bairro 4', cidade: 'Ipatinga' }
    ]
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('deve renderizar o cabeçalho corretamente', () => {
    const { getByText } = render(
      <DriverProfileScreen navigation={mockNavigation} />
    );
    
    expect(getByText('chevron-left')).toBeTruthy(); // Ícone de voltar
    expect(getByText('Ajuda')).toBeTruthy();
    expect(getByText('help-outline')).toBeTruthy(); // Ícone de ajuda
  });

  it('deve exibir todas as informações do motorista', () => {
    const { getByText } = render(
      <DriverProfileScreen navigation={mockNavigation} />
    );

    expect(getByText('John Doe')).toBeTruthy();
    expect(getByText('Passageiro')).toBeTruthy();
    expect(getByText('(31) 9 1234-5678')).toBeTruthy();
    expect(getByText('johndoe@gmail.com')).toBeTruthy();
  });

  it('deve exibir todos os endereços cadastrados', () => {
    const { getByText } = render(
      <DriverProfileScreen navigation={mockNavigation} />
    );

    expect(getByText('(Casa) R. 12, Bairro 1, Ipatinga')).toBeTruthy();
    expect(getByText('(Trabalho) R. 72, Bairro 4, Ipatinga')).toBeTruthy();
  });

  it('deve navegar para edição de perfil ao pressionar o botão', () => {
    const { getByText } = render(
      <DriverProfileScreen navigation={mockNavigation} />
    );

    fireEvent.press(getByText('Editar Perfil'));
    expect(mockNavigation.navigate).toHaveBeenCalledWith('EditProfile');
  });

  it('deve mostrar confirmação de logout ao pressionar Sair', () => {
    const { getByText } = render(
      <DriverProfileScreen navigation={mockNavigation} />
    );

    fireEvent.press(getByText('Sair'));
    expect(mockNavigation.setParams).toHaveBeenCalledWith({
      showLogoutConfirm: true
    });
  });

  it('deve voltar ao pressionar o ícone de voltar', () => {
    const { getByText } = render(
      <DriverProfileScreen navigation={mockNavigation} />
    );

    fireEvent.press(getByText('chevron-left'));
    expect(mockNavigation.goBack).toHaveBeenCalled();
  });

  it('deve renderizar o avatar do motorista', () => {
    const { getByText } = render(
      <DriverProfileScreen navigation={mockNavigation} />
    );

    expect(getByText('person')).toBeTruthy(); // Ícone de avatar
  });

  it('deve corresponder ao snapshot', () => {
    const { toJSON } = render(
      <DriverProfileScreen navigation={mockNavigation} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});