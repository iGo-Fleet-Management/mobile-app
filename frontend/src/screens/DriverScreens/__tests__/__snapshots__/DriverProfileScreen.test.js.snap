// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DriverProfileScreen deve corresponder ao snapshot 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 15,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "padding": 5,
        }
      }
    >
      <Text>
        chevron-left
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#007BFF",
            "marginRight": 5,
          }
        }
      >
        Ajuda
      </Text>
      <Text>
        help-outline
      </Text>
    </View>
  </View>,
  <RCTScrollView
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "marginVertical": 20,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#f5f5f5",
              "borderRadius": 50,
              "height": 100,
              "justifyContent": "center",
              "marginBottom": 15,
              "width": 100,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "borderRadius": 50,
                "height": 100,
                "justifyContent": "center",
                "width": 100,
              }
            }
          >
            <Text>
              person
            </Text>
          </View>
        </View>
        <Text
          style={
            {
              "fontSize": 22,
              "fontWeight": "bold",
            }
          }
        >
          John Doe
        </Text>
        <Text
          style={
            {
              "color": "#666",
              "fontSize": 16,
              "marginTop": 5,
            }
          }
        >
          Passageiro
        </Text>
      </View>
      <View
        style={
          {
            "marginBottom": 20,
            "paddingHorizontal": 20,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "bold",
              "marginBottom": 5,
              "marginTop": 15,
            }
          }
        >
          Idade:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 16,
            }
          }
        >
          20
        </Text>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "bold",
              "marginBottom": 5,
              "marginTop": 15,
            }
          }
        >
          Telefone:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 16,
            }
          }
        >
          (31) 9 1234-5678
        </Text>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "bold",
              "marginBottom": 5,
              "marginTop": 15,
            }
          }
        >
          Endere√ßos:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 16,
              "marginBottom": 5,
            }
          }
        >
          (
          Casa
          ) 
          R.
           
          12
          , 
          Bairro 1
          , 
          Ipatinga
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 16,
              "marginBottom": 5,
            }
          }
        >
          (
          Trabalho
          ) 
          R.
           
          72
          , 
          Bairro 4
          , 
          Ipatinga
        </Text>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "bold",
              "marginBottom": 5,
              "marginTop": 15,
            }
          }
        >
          Email:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 16,
            }
          }
        >
          johndoe@gmail.com
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#ddd",
            "borderRadius": 25,
            "borderWidth": 1,
            "marginBottom": 15,
            "marginHorizontal": 50,
            "opacity": 1,
            "paddingHorizontal": 20,
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Editar Perfil
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#ddd",
            "borderRadius": 25,
            "borderWidth": 1,
            "marginBottom": 30,
            "marginHorizontal": 50,
            "opacity": 1,
            "paddingHorizontal": 20,
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            {
              "color": "red",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Sair
        </Text>
      </View>
    </View>
  </RCTScrollView>,
]
`;
